audioipc/src/shm.rs

```diff
@@ -7,22 +7,96 @@ use crate::errors::*;
 use memmap::{Mmap, MmapMut, MmapOptions};
 use std::cell::UnsafeCell;
 use std::fs::{remove_file, File, OpenOptions};
-use std::path::Path;
+use std::path::PathBuf;
 use std::sync::{atomic, Arc};
+use std::convert::TryInto;
+use std::env::temp_dir;
+
+fn open_shm_file(id: &str) -> Result<File> {
+    if cfg!(target_os = "linux") {
+        let mut path = PathBuf::from("/dev/shm");
+        path.push(id);
+
+        if let Ok(file) = OpenOptions::new()
+            .read(true)
+            .write(true)
+            .create_new(true)
+            .open(&path) {
+                let _ = remove_file(&path);
+                return Ok(file);
+            }
+    };
+
+    let mut path = temp_dir();
+    path.push(id);
+
+    let file = OpenOptions::new()
+        .read(true)
+        .write(true)
+        .create_new(true)
+        .open(&path)?;
+
+    let _ = remove_file(&path);
+    Ok(file)
+}
+
+#[cfg(unix)]
+fn allocate_file(file: &File, size: usize) -> Result<()> {
+    use std::os::unix::io::AsRawFd;
+
+    // Try Linux-specific fallocate.
+    #[cfg(target_os = "linux")]
+    unsafe {
+        if libc::fallocate(file.as_raw_fd(), 0, 0, size.try_into().unwrap()) == 0 {
+            return Ok(())
```

Comment:
I see `size.try_into().unwrap()` is used in several places. Maybe it can be declared at first of `allocate_file` like `let size = u64::try_from(size).unwrap()` and then use `size` in the following code.

Reply:
It's not the same type everywhere, unfortunately.  fallocate and fcntl (via fstore_t) takes an off_t, set_len takes a u64.  off_t is usually an i64 these days, but it can be i32 on older 32-bit systems.  So converting in-place is uglier, but we get the correct type everywhere.